### PR TITLE
Redecoration improvements

### PR DESCRIPTION
**Pull request by parar020100**
- Redecoration switch enables self-repair
- If no interior selected, a random one is picked
- A simultaneous demat/mat animation instead of just "remove+appear"
- The new TARDIS inherits the properties of the old one: flight, float, hads, spindir, cloak, handbrake, physlock, isomorphic, vortex flight and fast return coords

- Added optional skin randomization at spawn, extensions can select the excluded skins

### CODE REVIEW

File: lua/entities/gmod_tardis/modules/sh_health.lua [hunk 1]
```diff
@@ -151,16 +151,96 @@ if SERVER then
...
+					net.WriteEntity(ent)
+				net.Send(owner)
+			end
+
+
+			-- fly away
+			self:SetData("redecorate-demattime", CurTime())
+			self:SetFastRemat(true)
+			self:SetPhyslock(true) -- should not move
+			ent:SetPhyslock(true)
```
Comment (MattJeanes) [inline lua/entities/gmod_tardis/modules/sh_health.lua:240]: This results in a weird experience, is this strictly necessary especially with the parenting stuff added in now?

File: lua/entities/gmod_tardis/modules/sh_health.lua [hunk 2]
```diff
@@ -174,6 +254,57 @@ if SERVER then
...
+	ENT:AddHook("StopMat", "redecorate_restore_data", function(self)
+		if self:GetData("parent-saved-data") then
+
+			local saved_data = self:GetData("parent-saved-data")
+
+			for k,v in ipairs(migrated_data_names) do
+				self:SetData(v, saved_data[v], true)
+			end
+			if IsValid(self.interior) then
+				self.interior:SetData("security", self:GetData("parent-security", false), true)
```
Comment (MattJeanes) [inline lua/entities/gmod_tardis/modules/sh_health.lua:296]: Would it make more sense to have an interior_migrated_data_names or something here for this?

**Review by AmyJeanes — COMMENTED** [review]
Looking really nice and in game too working nicely for the most part! Just have a couple questions before approving

**Review by AmyJeanes — COMMENTED** [review]
Looking pretty good now, it's a shame about the physlock but fair enough. Speaking of which though, the physlock is always enabled after the redecoration finishes now which I assume is not intended behaviour.

**Comment by parar020100** [comment]
> Looking pretty good now, it's a shame about the physlock but fair enough. Speaking of which though, the physlock is always enabled after the redecoration finishes now which I assume is not intended behaviour.

Should be fixed now

File: lua/entities/gmod_tardis/modules/sh_timer.lua [hunk 3]
```diff
@@ -0,0 +1,29 @@
+local timers = {}
```
Comment (MattJeanes) [inline lua/entities/gmod_tardis/modules/sh_timer.lua:1]: This should definitely be initialised as self.timers on an Initialize hook (or maybe pre-initialize) or this will conflict with other TARDISes

File: lua/entities/gmod_tardis_interior/modules/sh_timer.lua [hunk 4]
```diff
@@ -0,0 +1,29 @@
+local timers = {}
```
Comment (MattJeanes) [inline lua/entities/gmod_tardis_interior/modules/sh_timer.lua:1]: Ditto on self.timers

**Review by AmyJeanes — CHANGES_REQUESTED** [review]
Behaviour in game is feeling better now and code looks improved, although some new fairly minor issues found in code + behaviour.

File: lua/entities/gmod_tardis_interior/modules/sh_timer.lua (hunk 4 above)
Comment (parar020100) [reply lua/entities/gmod_tardis_interior/modules/sh_timer.lua]: Fixed

File: lua/entities/gmod_tardis/modules/sh_timer.lua (hunk 3 above)
Comment (parar020100) [reply lua/entities/gmod_tardis/modules/sh_timer.lua]: Fixed

Comment (parar020100) [reply lua/entities/gmod_tardis/modules/sh_timer.lua]: `ENT.timers` does work though (without any hooks)
I've checked that different TARDISes get separate hooks now

File: lua/entities/gmod_tardis/modules/sh_redecorate.lua [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+		if not IsValid(child) then
+			error("Redecoration failed: failed to spawn the replacement")
+		end
+		return IsValid(child) -- if gone wrong, finish repair
+	end
+
+	ENT:AddHook("StopDemat", "redecorate_remove_parent", function(self)
+		if self:GetData("redecorate") then
+			self:GetData("redecorate_child"):SetData("redecorate_parent", nil, true)
```
[the remark this answers is not in the document]
Comment (parar020100) [reply lua/entities/gmod_tardis/modules/sh_redecorate.lua]: +

File: lua/entities/gmod_tardis/modules/sh_redecorate.lua [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+		local child = TARDIS:SpawnTARDIS(ply, {
+			pos = self:GetPos(),
+			metadataID = self:GetData("redecorate-interior") or "default",
+			redecorate_parent = self,
+			ext_data = ext_saved_data,
+			int_data = int_saved_data,
+		})
+
+		if not IsValid(child) then
+			error("Redecoration failed: failed to spawn the replacement")
```
[the remark this answers is not in the document]
Comment (parar020100) [reply lua/entities/gmod_tardis/modules/sh_redecorate.lua]: Fixed

File: lua/entities/gmod_tardis/modules/sh_redecorate.lua [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+			self:Remove()
+		end
+	end)
+
+	ENT:AddHook("MatStart", "redecorate_sync", function(self)
+		local parent = self:GetData("redecorate_parent")
+		if not parent then return end
+
+		self:SetCollisionGroup(COLLISION_GROUP_WORLD)
+		self:SetPos(parent:GetPos())
```
[the remark this answers is not in the document]
Comment (parar020100) [reply lua/entities/gmod_tardis/modules/sh_redecorate.lua:92]: I've added safeguards to prevent that one in a25844d5f13117fd639e2c45d1f69c777d3d8622
Now, when the parent is dropped, it cannot be picked up again (COLLISION_GROUP_IN_VEHICLE)

File: lua/entities/gmod_tardis/modules/sh_redecorate.lua [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+		apply_saved_data(self, ext_saved_data, ext_saved_data_names)
+		apply_saved_data(self.interior, int_saved_data, ext_saved_data_names)
+
+		self:SetData("redecorate_parent_ext_data", nil, true)
+		self:SetData("redecorate_parent_int_data", nil, true)
+
+		constraint.RemoveAll(parent) -- drop everything attached
+		parent:SetFastRemat(true)
+		parent:ForcePlayerDrop()
```
[the remark this answers is not in the document]
Comment (parar020100) [reply lua/entities/gmod_tardis/modules/sh_redecorate.lua]: I've added safeguards to prevent that one in a25844d5f13117fd639e2c45d1f69c777d3d8622
Now, when the parent is dropped, it cannot be picked up or moved again (COLLISION_GROUP_IN_VEHICLE)
+ it won't move by itself due to physlock

File: lua/entities/gmod_tardis/modules/sh_redecorate.lua [hunk 9]
```diff
@@ -22,17 +22,13 @@ local int_saved_data_names = {
 	"security"
 }
 
-local function apply_saved_data(object, data, names)
+function TARDIS:apply_redecoration_data(object, data)
```
Comment (MattJeanes) [inline lua/entities/gmod_tardis/modules/sh_redecorate.lua:25]: A local function seemed fine here, why was this changed? Also we do not want to generally define TARDIS globals inside the entity folder

File: lua/entities/gmod_tardis/modules/sh_redecorate.lua [hunk 10]
```diff
@@ -68,14 +64,17 @@ if SERVER then
...
+			ErrorNoHalt("Redecoration failed: failed to spawn the replacement")
 		end
 		return IsValid(child) -- if gone wrong, finish repair
 	end
 
 	ENT:AddHook("StopDemat", "redecorate_remove_parent", function(self)
 		if self:GetData("redecorate") then
-			self:GetData("redecorate_child"):SetData("redecorate_parent", nil, true)
+			local child = self:GetData("redecorate_child")
+			if child then
```
Comment (MattJeanes) [inline lua/entities/gmod_tardis/modules/sh_redecorate.lua:75]: Check IsValid here not just if the variable exists

File: lua/entities/gmod_tardis/modules/sh_timer.lua [hunk 11]
```diff
@@ -1,23 +1,24 @@
-local timers = {}
+ENT.timers = {}
```
Comment (MattJeanes) [inline lua/entities/gmod_tardis/modules/sh_timer.lua:1]: This is still global, we need to create this on self inside an Initialize hook

File: lua/entities/gmod_tardis_interior/modules/sh_timer.lua [hunk 12]
```diff
@@ -1,23 +1,23 @@
-local timers = {}
+ENT.timers = {}
```
Comment (MattJeanes) [inline lua/entities/gmod_tardis_interior/modules/sh_timer.lua:1]: ditto on timers

**Review by AmyJeanes — COMMENTED** [review]
In-game behaviour looks good now, just a couple of little notes and then I think we're ready to go!

**Review by AmyJeanes — APPROVED** [review]
